core/transformation/utils.py

```diff
@@ -114,3 +115,18 @@ def extract_postcodes(s: str) -> list[str] | None:
         if postcode_area_matches == []:
             return None
     return postcode_area_matches
+
+
+def create_dataframe(data: dict[str, Iterable]) -> pd.DataFrame:
```

Comment:
is this type hinting too loose? Can this be union of the specific types expected rather than all Python iterables?

Reply:
Since this theoretically could take a string as a param, and give all sorts of weird results.

Reply:
I'd originally considered `Union[pd.Series, List]` but then realised we have a tuple coming in as well, and that point I thought it best to simply type-hint `Iterable` rather than a three-way `Union` to be clean. I reckon it's okay to keep type hint as-is based on the [Pandas documentation for `pd.Series`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html) which suggests that the constructor data argument can be "array-like, Iterable, dict, or scalar value"

Reply:
since python 3.10 you can union types with |

Reply:
so I _think_ it would look something like
`dict[str, pd.Series | List]`

Reply:
what is the tuple that comes in exactly?

Reply:
I've just pushed a commit converting from `Union` syntax to pipe operator. I've kept tuple in the definition for now because the tuple examples come from the below code:

`questions, answers = zip(*[(q, a) for q, a in zip(questions, answers) if pd.notna(a)])
  return create_dataframe(
        {
            "Programme ID": [programme_id] * len(questions),
            "Question": questions,
            "Answer": answers,
        }
    )`

To convert back from tuples to lists, we'd have to call `list(questions)` and `list(answers)` and this doesn't feel worth it just for the sake of the type hint?

Reply:
In terms of clean code and re-useability, making your functions less variable/ambiguous should be a priority - It makes it easier to re-use / extend in future and reduces the risk of misuse. Also, if we used static type checker like MyPy it would probably complain about this. However, it's probably fine here as built as part of a bespoke pipeline. 

But, please do add it do the docstring description, otherwise it's not clear what kind of Tuple is actually expected.

Reply:
I've parameterised the `list` and `tuple` type hints to make the type-hints self-documenting, rather than adding extra text to the docstring, such that the function interface now looks like:

```
def create_dataframe(
    data: dict[str, pd.Series | list[datetime | int | str] | tuple[datetime | int | str]]
) -> pd.DataFrame:
```